### PR TITLE
GobEncode: avoid reallocating initial slice

### PR DESCRIPTION
**Pull request by serprex**

### CODE REVIEW

**Review by mwoss — APPROVED** [review]
Thanks for your contribution @serprex :D

File: decimal.go [hunk 1]
```diff
@@ -1506,19 +1506,18 @@ func (d *Decimal) UnmarshalBinary(data []byte) error {
...
-	if v2, err = d.value.GobEncode(); err != nil {
-		return
+	// exp is written first, but encode value first to know output size
+	var valueData []byte
+	if valueData, err = d.value.GobEncode(); err != nil {
+		return nil, err
 	}
 
+	// Write the exponent in front since it's a fixed size
+	expData := make([]byte, 4, len(valueData)+4)
```
[the remark this answers is not in the document]
Comment (serprex) [reply decimal.go:1516]: Yes, that capacity is final length

File: decimal.go [hunk 2]
```diff
@@ -1506,19 +1506,18 @@ func (d *Decimal) UnmarshalBinary(data []byte) error {
...
-	var v2 []byte
-	if v2, err = d.value.GobEncode(); err != nil {
-		return
+	// exp is written first, but encode value first to know output size
+	var valueData []byte
+	if valueData, err = d.value.GobEncode(); err != nil {
+		return nil, err
 	}
 
+	// Write the exponent in front since it's a fixed size
```
[the remark this answers is not in the document]
Comment (serprex) [reply decimal.go]: fixed

File: decimal.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mwoss) [reply decimal.go:1516]: Great, thanks for confirming :3

**Comment by mwoss** [comment]
LGTM! Ready for merge :3

**Pull request merged**